### PR TITLE
feat(analyzer): top level suppression

### PR DESCRIPTION
**Pull request by ematipico**
## Summary

Closes https://github.com/biomejs/biome/issues/4305

This PR introduces a new type of suppression: top-level suppression.

### TL;DR: 

```js
/** biome-ignore lint/style/useConst: reason */

let foo = 1;
let bar = 12;
```

The top-level suppression works only if:
- it's at the beginning of the file
- there are two newlines after the comment

This means this comment is seen as a "next line suppression", and it should raise a diagnostic for `let bar = 12`:

```js
/** biome-ignore lint/style/useConst: reason */
let foo = 1;
let bar = 12;
```

To recognise them, the `parse_suppression_comment` now accepts a `SyntaxToken`. If the range of the comment starts at `0` and there are exactly two newlines after the comment, it means that it's a top-level suppression. The range check is done with the trivia, because we want to make sure to catch the `0` position. 

This PR also removes legacy suppression comments.

### More changes

The `Rule` type generates the suppression via `R::suppress` function. Initially, I wanted to change the type returned to a `Vec`, but that doesn't play well with our infrastructure because the returned type is mapped exactly to a code action, which means that returning a vector wasn't possible.

I renamed `R::suppress` to `R::inline_suppression`, and I created a new function called `R::toplevel_suppression`. This function leverage the already existing trait `SuppressionAction`, which now requires to implement a new function called `apply_top_level_suppression`. The logic of this function is way simpler than that of inline suppression. This function accepts the first token of a root CST, and from there, it's very simple to prepend a new comment trivia.

I also created two new *known* variants to the action category called `OtherActionCategory`, where we can add new known LSP client signals. These two signals were added to the LSP client capabilities. 

## Test Plan

 I added various tests for the CLI and the analyzer.

I also updated the LSP tests, which now send a new action for the top-level suppression. I manually tested this with Zed, and I can correctly see the new top-level action pulled and applied via the editor.

### CODE REVIEW

**Comment by vohoanglong0107** [comment]
May I ask why we need to enforce top level suppression to be a block comment? Some languages, like GraphQL 😅, only have single line comment.
Implementation wise I don't think it would impact other languages that much, but I think it's worth mentioning in the Changelog

**Comment by ematipico** [comment]
> why we need to enforce top level suppression to be a block comment? Some languages, like GraphQL 😅, only have single line comment.

I wasn't aware of that. I should be able to lift the restriction

**Comment by ematipico** [comment]
@vohoanglong0107 The algorithm has been updated, and the heuristic is different. I updated the PR description to reflect that.

**Review by vohoanglong0107 — APPROVED** [review]

File: crates/biome_js_analyze/tests/suppression/correctness/noUnusedVariables/ts-module-export.ts
Comment (ematipico) [inline crates/biome_js_analyze/tests/suppression/correctness/noUnusedVariables/ts-module-export.ts:1]: This wasn't needed, it didn't emit any suppressions

File: crates/biome_analyze/src/diagnostics.rs [hunk 1]
```diff
@@ -164,15 +167,33 @@ impl SuppressionDiagnostic {
             range,
             message: message.to_string(),
             tags: DiagnosticTags::empty(),
+            advice: SuppressionAdvice::default(),
         }
     }
 
-    pub(crate) fn with_tags(mut self, tags: DiagnosticTags) -> Self {
-        self.tags |= tags;
+    pub(crate) fn note(mut self, message: impl Into<String>, range: impl Into<TextRange>) -> Self {
```
Comment (ematipico) [inline crates/biome_analyze/src/diagnostics.rs:174]: This new API was created to suppose this new diagnostic: https://github.com/biomejs/biome/pull/4306/files#diff-6b7a758a5c6e91dd2722ccc423ad136ff4e811311c0d49b80c659d61cc54e2a6

Neat :)

File: crates/biome_analyze/src/lib.rs [hunk 2]
```diff
@@ -238,12 +281,14 @@ struct LineSuppression {
     suppress_all: bool,
     /// List of all the rules this comment has started suppressing (must be
     /// removed from the suppressed set on expiration)
-    suppressed_rules: Vec<RuleFilter<'static>>,
+    suppressed_rules: FxHashSet<RuleFilter<'static>>,
     /// List of all the rule instances this comment has started suppressing.
-    suppressed_instances: Vec<(RuleFilter<'static>, String)>,
+    suppressed_instances: FxHashMap<String, RuleFilter<'static>>,
```
Comment (ematipico) [inline crates/biome_analyze/src/lib.rs:286]: I had to swap this, because the primary key is the instance, not the filter.

File: crates/biome_analyze/src/lib.rs [hunk 3]
```diff
@@ -527,21 +593,9 @@ where
...
-                )
-                .with_tags(DiagnosticTags::DEPRECATED_CODE)
-            });
-
-            let signal = signal
-                .with_action(|| update_suppression(self.root, token, is_leading, index, text));
-
-            (self.emit_signal)(&signal)?;
+        if is_top_level_suppression {
+            self.top_level_suppressions.expand_range(range);
```
Comment (ematipico) [inline crates/biome_analyze/src/lib.rs:597]: We might encounter multiple top level comments, so we need to expand its initial range until the end.

File: crates/biome_analyze/src/lib.rs [hunk 4]
```diff
@@ -635,75 +694,27 @@ fn range_match(filter: Option<TextRange>, range: TextRange) -> bool {
...
+    // TODO: docs
+    TopLevel(&'a str),
     /// A suppression disabling all lints eg. `// biome-ignore lint`
     Everything,
     /// A suppression disabling a specific rule eg. `// biome-ignore lint/style/useWhile`
     Rule(&'a str),
     /// A suppression to be evaluated by a specific rule eg. `// biome-ignore lint/correctness/useExhaustiveDependencies(foo)`
     RuleInstance(&'a str, &'a str),
-    /// A suppression using the legacy syntax to disable a specific rule eg. `// biome-ignore lint(style/useWhile)`
-    MaybeLegacy(&'a str),
```
Comment (ematipico) [inline crates/biome_analyze/src/lib.rs:651]: Not needed anymore. It was an old rome thing (the TS codebase one)

File: crates/biome_analyze/src/signals.rs [hunk 5]
```diff
@@ -115,7 +117,15 @@ pub struct AnalyzerAction<L: Language> {
...
-        self.category.matches(SUPPRESSION_ACTION_CATEGORY)
+        self.is_inline_suppression() || self.is_top_level_suppression()
+    }
+
+    pub fn is_inline_suppression(&self) -> bool {
+        self.category.matches(SUPPRESSION_INLINE_ACTION_CATEGORY)
+    }
+
+    pub fn is_top_level_suppression(&self) -> bool {
+        self.category.matches(SUPPRESSION_TOP_LEVEL_ACTION_CATEGORY)
```
Comment (ematipico) [inline crates/biome_analyze/src/signals.rs:128]: We will need these two new functions for the upcoming feature from @anthonyshew

**Review by arendjr — APPROVED** [review]
Great stuff! I think people are going to love this!

I do have my doubts about the requirement to have two newlines following the suppression comment. I understand why you’re doing it, but it’s not intuitive and hard to explain. It feels like a “magic behavior” that you have to know about that will easily bite you if you don’t.

Can’t we have a syntax like `biome-ignore-all` so that we can lift this requirement?

**Comment by ematipico** [comment]
> Great stuff! I think people are going to love this!
> 
> I do have my doubts about the requirement to have two newlines following the suppression comment. I understand why you’re doing it, but it’s not intuitive and hard to explain. It feels like a “magic behavior” that you have to know about that will easily bite you if you don’t.
> 
> Can’t we have a syntax like `biome-ignore-all` so that we can lift this requirement?

I put great thoughts into what you suggested. At the beginning, I thought it could work, however I found two downsides that are worth considering.

First, lifting the "2 newlines" requirement could cause issues with the import sorting, where a comment is moved with its relative import statement:

```js
// biome-ignore-all lint/style/useConst: reason
import zod from "zod"
import lodash from "lodash" // added by IDE automatically 
```

If we execute the first import statement will be moved to the end, together with the suppression comment. 

Second, we need to consider if this new suppression kind has repercussions on the formatter. If so, how? 

So, if we would want to implement a new suppression kind, we would need to consider what we need to change, where and how.

**Comment by arendjr** [comment]
Yeah, the import sorting is a valid concern indeed. We could solve it by adding an exception in the sorter, which is one of the benefits of an integrated solution like Biome. It would be extra effort for us, but I think it may benefit users.

Not sure what impact it would have on the formatter though. At least I can’t really think of something… The formatter should leave comments where they are, so it should remain a top-level suppression, right?

I guess if you wanted to you could let the formatter recognize `biome-ignore-all` comments  and insert two newlines after. But given it wouldn’t be a requirement anymore I also don’t see a really good reason to do so.

**Comment by ematipico** [comment]
OK, I will apply this change in a later PR.

cc @Conaclos, as you're involved in the import sorting, and you should be interested in this conversation.
